aaelf64-morello/aaelf64-morello.rst

```diff
@@ -827,3 +844,91 @@ veneer is used. The BX changes the execution state from A64 to C64:
   adrp c16, sym
   add c16, c16, :lo12:sym
   br c16
+
+TLS for the pure capability ABI
+-------------------------------
+
+The design is based on TLSDESC, with the purpose of minimizing the performance
+differences between A64 and C64, while providing strict bounds when resolving
+TLS globals. Only the General Dynamic access model and the relaxation from
+General Dynamic to Local Exec are currently specified. However, this leaves the
+opportunity to add support for the other access models in the future.
+
+TLS static block
+^^^^^^^^^^^^^^^^
+
+The static block layout is the same used in AArch64 (Variant 1, see [TLS_]), with
+the only exception that TCB and the DTV pointer are capabilities.
+
+Thread pointer
+^^^^^^^^^^^^^^
+
+The thread pointer is a capability, held in ``CTPIDR_EL0``. The thread pointer
+needs to have the read, write, read capability and write capability permissions
+and bounds such that the entire TLS static block is accessible.
+
+Resolver functions
+^^^^^^^^^^^^^^^^^^
+
+A resolver function takes arguments in c0 (address of the TLS GOT slot), and c2
+(a copy of the thread pointer) and returns a pointer to the TLS global in c0
+and its size in x1. The resolver function has a custom calling convention that
+must preserve all registers except c0 and c1.
+
+Considerations:
+
+- Any dynamically loaded modules will be placed outside of the bounds of the
+  thread pointer, so a resolver function cannot return an offset from the
+  thread pointer, but rather needs to return a pointer (capability).
+- To minimize reading of ``CTPIDR_EL0``, the resolver functions take a
+  copy of ``CTPIDR_EL0`` as an argument and preserve it.
+- Not setting the bounds directly in the resolver allows shorter sequences
+  for the General Dynamic access model.
+
+
+Static TLS block resolver
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If the TLS variable is in the static block, while resolving the
+``R_MORELLO_TLSDESC`` relocation, the dynamic linker will place in the two GOT
+slots associated with this variable:
+
+- A capability to the static TLS block resolver function at offset 0.
+- The offset of the variable in the static TLS block at offset 16 (8 bytes).
+- The size of the variable at offset 24 (8 bytes).
```

Comment:
Isn't that an implementation detail of the runtime you're using, not part of the ABI? AFAIK the only requirement is that the first capability point to the resolver?